client/python/openlineage/client/generated/base.py

```diff
@@ -123,7 +123,7 @@ class Dataset(RedactMixin):
     name: str
     """The unique name for that dataset within that namespace"""
 
-    facets: dict[str, DatasetFacet] | None = field(factory=dict, kw_only=True)  # type: ignore[assignment]
+    facets: Optional[dict[str, DatasetFacet]] = field(factory=dict, kw_only=True)
```

Comment:
But why?

Reply:
`from __future__ import annotations` enable this support on Python 3.7+

Reply:
it doesn't solve this, we already have this on top of base.py file

here is full code to reproduce

```
from __future__ import annotations

from openlineage.client.generated.base import RunEvent

from json import loads

_data = """
  {
    "eventTime": "2023-12-16T14:22:11.949Z",
    "producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
    "schemaURL": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/RunEvent",
    "eventType": "COMPLETE",
    "run": {
      "runId": "59fc8906-4a4a-45ab-9a54-9cc2d399e10e",
      "facets": {
        "parent": {
          "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
          "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/ParentRunFacet.json#/$defs/ParentRunFacet",
          "run": {
            "runId": "daf8bcc1-cc3c-41bb-9251-334cacf698fa"
          },
          "job": {
            "namespace": "TESTSchedulerID",
            "name": "TESTParentJobName4"
          }
        },
        "spark.logicalPlan": {
          "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
          "_schemaURL": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/RunFacet",
          "plan": [
            {
              "class": "org.apache.spark.sql.execution.command.CreateDataSourceTableAsSelectCommand",
              "num-children": 1,
              "table": {
                "product-class": "org.apache.spark.sql.catalyst.catalog.CatalogTable",
                "identifier": {
                  "product-class": "org.apache.spark.sql.catalyst.TableIdentifier",
                  "table": "dst_table_test_from_src_table_df",
                  "database": "test_db"
                },
                "tableType": {
                  "product-class": "org.apache.spark.sql.catalyst.catalog.CatalogTableType",
                  "name": "MANAGED"
                },
                "storage": {
                  "product-class": "org.apache.spark.sql.catalyst.catalog.CatalogStorageFormat",
                  "compressed": false,
                  "properties": null
                },
                "schema": {
                  "type": "struct",
                  "fields": []
                },
                "provider": "parquet",
                "partitionColumnNames": [],
                "owner": "",
                "createTime": 1702736481797,
                "lastAccessTime": -1,
                "createVersion": "",
                "properties": null,
                "unsupportedFeatures": [],
                "tracksPartitionsInCatalog": false,
                "schemaPreservesCase": true,
                "ignoredProperties": null
              },
              "mode": null,
              "query": 0,
              "outputColumnNames": "[id, randomid, zip]"
            },
            {
              "class": "org.apache.spark.sql.execution.datasources.LogicalRelation",
              "num-children": 0,
              "relation": null,
              "output": [
                [
                  {
                    "class": "org.apache.spark.sql.catalyst.expressions.AttributeReference",
                    "num-children": 0,
                    "name": "id",
                    "dataType": "long",
                    "nullable": true,
                    "metadata": {},
                    "exprId": {
                      "product-class": "org.apache.spark.sql.catalyst.expressions.ExprId",
                      "id": 9,
                      "jvmId": "78343417-b80d-45a2-b489-e7d8920e61dd"
                    },
                    "qualifier": []
                  }
                ],
                [
                  {
                    "class": "org.apache.spark.sql.catalyst.expressions.AttributeReference",
                    "num-children": 0,
                    "name": "randomid",
                    "dataType": "string",
                    "nullable": true,
                    "metadata": {},
                    "exprId": {
                      "product-class": "org.apache.spark.sql.catalyst.expressions.ExprId",
                      "id": 10,
                      "jvmId": "78343417-b80d-45a2-b489-e7d8920e61dd"
                    },
                    "qualifier": []
                  }
                ],
                [
                  {
                    "class": "org.apache.spark.sql.catalyst.expressions.AttributeReference",
                    "num-children": 0,
                    "name": "zip",
                    "dataType": "string",
                    "nullable": true,
                    "metadata": {},
                    "exprId": {
                      "product-class": "org.apache.spark.sql.catalyst.expressions.ExprId",
                      "id": 11,
                      "jvmId": "78343417-b80d-45a2-b489-e7d8920e61dd"
                    },
                    "qualifier": []
                  }
                ]
              ],
              "catalogTable": {
                "product-class": "org.apache.spark.sql.catalyst.catalog.CatalogTable",
                "identifier": {
                  "product-class": "org.apache.spark.sql.catalyst.TableIdentifier",
                  "table": "src_table_test",
                  "database": "test_db"
                },
                "tableType": {
                  "product-class": "org.apache.spark.sql.catalyst.catalog.CatalogTableType",
                  "name": "MANAGED"
                },
                "storage": {
                  "product-class": "org.apache.spark.sql.catalyst.catalog.CatalogStorageFormat",
                  "locationUri": null,
                  "inputFormat": "org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat",
                  "outputFormat": "org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat",
                  "serde": "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe",
                  "compressed": false,
                  "properties": null
                },
                "schema": {
                  "type": "struct",
                  "fields": [
                    {
                      "name": "id",
                      "type": "long",
                      "nullable": true,
                      "metadata": {}
                    },
                    {
                      "name": "randomid",
                      "type": "string",
                      "nullable": true,
                      "metadata": {}
                    },
                    {
                      "name": "zip",
                      "type": "string",
                      "nullable": true,
                      "metadata": {
                        "__CHAR_VARCHAR_TYPE_STRING": "varchar(10)"
                      }
                    }
                  ]
                },
                "provider": "parquet",
                "partitionColumnNames": [],
                "owner": "test_user@ad.test.net",
                "createTime": 1682523315000,
                "lastAccessTime": 0,
                "createVersion": "3.3.1",
                "properties": null,
                "stats": null,
                "unsupportedFeatures": [],
                "tracksPartitionsInCatalog": false,
                "schemaPreservesCase": true,
                "ignoredProperties": null
              },
              "isStreaming": false
            }
          ]
        },
        "spark_version": {
          "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
          "_schemaURL": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/RunFacet",
          "spark-version": "3.3.1",
          "openlineage-spark-version": "1.5.0"
        },
        "processing_engine": {
          "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
          "_schemaURL": "https://openlineage.io/spec/facets/1-1-0/ProcessingEngineRunFacet.json#/$defs/ProcessingEngineRunFacet",
          "version": "3.3.1",
          "name": "spark",
          "openlineageAdapterVersion": "1.5.0"
        },
        "environment-properties": {
          "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
          "_schemaURL": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/RunFacet",
          "environment-properties": {}
        }
      }
    },
    "job": {
      "namespace": "TESTSchedulerID",
      "name": "test_user_spark.execute_create_data_source_table_as_select_command.dst_table_test_from_src_table_df",
      "facets": {}
    },
    "inputs": [
      {
        "namespace": "s3a://test_db-db",
        "name": "src_table_test",
        "facets": {
          "dataSource": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/DatasourceDatasetFacet.json#/$defs/DatasourceDatasetFacet",
            "name": "s3a://test_db-db",
            "uri": "s3a://test_db-db"
          },
          "schema": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/SchemaDatasetFacet.json#/$defs/SchemaDatasetFacet",
            "fields": [
              {
                "name": "id",
                "type": "long"
              },
              {
                "name": "randomid",
                "type": "string"
              },
              {
                "name": "zip",
                "type": "string"
              }
            ]
          },
          "symlinks": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/SymlinksDatasetFacet.json#/$defs/SymlinksDatasetFacet",
            "identifiers": [
              {
                "namespace": "hive://hive-metastore.hive.svc.cluster.local:9083",
                "name": "shopify.raw_product_catalog",
                "type": "TABLE"
              }
            ]
          }
        },
        "inputFacets": {}
      }
    ],
    "outputs": [
      {
        "namespace": "s3a://test_db-db",
        "name": "dst_table_test_from_src_table_df",
        "facets": {
          "dataSource": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/DatasourceDatasetFacet.json#/$defs/DatasourceDatasetFacet",
            "name": "s3a://test_db-db",
            "uri": "s3a://test_db-db"
          },
          "schema": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/SchemaDatasetFacet.json#/$defs/SchemaDatasetFacet",
            "fields": [
              {
                "name": "id",
                "type": "long"
              },
              {
                "name": "randomid",
                "type": "string"
              },
              {
                "name": "zip",
                "type": "string"
              }
            ]
          },
          "columnLineage": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-1/ColumnLineageDatasetFacet.json#/$defs/ColumnLineageDatasetFacet",
            "fields": {
              "id": {
                "inputFields": [
                  {
                    "namespace": "s3a://test_db-db",
                    "name": "/src_table_test",
                    "field": "comments"
                  }
                ]
              },
              "randomid": {
                "inputFields": [
                  {
                    "namespace": "s3a://test_db-db",
                    "name": "/src_table_test",
                    "field": "products"
                  }
                ]
              },
              "zip": {
                "inputFields": [
                  {
                    "namespace": "s3a://test_db-db",
                    "name": "/src_table_test",
                    "field": "platform"
                  }
                ]
              }
            }
          },
          "symlinks": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/SymlinksDatasetFacet.json#/$defs/SymlinksDatasetFacet",
            "identifiers": [
              {
                "namespace": "hive://hive-metastore.hive.svc.cluster.local:9083",
                "name": "shopify.fact_order_new5",
                "type": "TABLE"
              }
            ]
          },
          "lifecycleStateChange": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/LifecycleStateChangeDatasetFacet.json#/$defs/LifecycleStateChangeDatasetFacet",
            "lifecycleStateChange": "CREATE"
          }
        },
        "outputFacets": {}
      }
    ]
  }
"""

_data = loads(_data)

a = RunEvent(**_data)

import cattr
x = cattr.Converter().structure(_data, RunEvent)

print(x)
```

Reply:
Ok, cattrs probably using `typing.get_type_hints` which uses `eval` under the hood, and `X | Y` syntax on Python < 3.10 is not a valid code.

Reply:
The code actually does not work properly, it doesn't throw errors after the change though. E.g. facets are described as `DatasetFacet` without additional properties.

Reply:
It's difficult to have fully working example :(
Do you know if following problem might be resolved with pydantic?

```json
{
    "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
    "_schemaURL": "https://openlineage.io/spec/facets/1-0-1/ColumnLineageDatasetFacet.json#/$defs/ColumnLineageDatasetFacet",
    "fields": {
        "id": {
            "inputFields": [
                {
                    "namespace": "s3a://test_db-db",
                    "name": "/src_table_test",
                    "field": "comments"
                }
            ]
        },
        "randomid": {
            "inputFields": [
                {
                    "namespace": "s3a://test_db-db",
                    "name": "/src_table_test",
                    "field": "products"
                }
            ]
        },
        "zip": {
            "inputFields": [
                {
                    "namespace": "s3a://test_db-db",
                    "name": "/src_table_test",
                    "field": "platform"
                }
            ]
        }
    }
}
```

but we're given:
```python
@define
class ColumnLineageDatasetFacet(DatasetFacet):
    fields: dict[str, Fields]
```
and
```python
@define
class SchemaDatasetFacet(DatasetFacet):
    fields: Optional[list[SchemaDatasetFacetFields]] = field(factory=list)  # type: ignore[assignment]
    """The fields of the data source."""
```

cattrs is looking at unique attribute per each class. In this case it fails and I don't see any option to first try to structure nested field and see what class it could be.

One option would be to somehow use keys under which facets are located, e.g. `SQLJobFacet` is located under `sql` key. Probably this is how pydantic handles things?

Reply:
> One option would be to somehow use keys under which facets are located, e.g. SQLJobFacet is located under sql key. Probably this is how pydantic handles things?

pydantic is designed to deserialize/serialize classes, and it is expected that there will be some class with `sql: SQLJobFacet` attribute, to parse input JSON to specified class.

This class can be [TypedDict](https://docs.pydantic.dev/2.3/usage/types/dicts_mapping/#typeddict), allowing both parsing specific keys as specific class, and later accessing values as dict elements rather than object attributes. But unlike `dict[K, V]` it lacks of annotation for extra values in this dict, telling pydantic that unknown keys should be parsed as generic `DatasetFacet` or `RunFacet`.
This probably can be achieved by some custom validator.

Reply:
@dolfinus could you provide small example with minimal set of classes just to show how you would see this?

Reply:
but current implementation doesn't touch pydantic at all no?

Reply:
we could consider pydantic if it solves all problems with deserialization

Reply:
I have made very preliminary approach for this, it's available here https://github.com/mgorsk1/OpenLineage/pull/1/files

and deserialization works out of the box:

```
from json import loads
from openlineage.client.generated.base import RunEvent
from deepdiff import DeepDiff

_data = """
  {
    "eventTime": "2023-12-16T14:22:11.949Z",
    "producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
    "schemaURL": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/RunEvent",
    "eventType": "COMPLETE",
    "run": {
      "runId": "59fc8906-4a4a-45ab-9a54-9cc2d399e10e",
      "facets": {
        "parent": {
          "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
          "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/ParentRunFacet.json#/$defs/ParentRunFacet",
          "run": {
            "runId": "daf8bcc1-cc3c-41bb-9251-334cacf698fa"
          },
          "job": {
            "namespace": "TESTSchedulerID",
            "name": "TESTParentJobName4"
          }
        },
        "spark.logicalPlan": {
          "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
          "_schemaURL": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/RunFacet",
          "plan": [
            {
              "class": "org.apache.spark.sql.execution.command.CreateDataSourceTableAsSelectCommand",
              "num-children": 1,
              "table": {
                "product-class": "org.apache.spark.sql.catalyst.catalog.CatalogTable",
                "identifier": {
                  "product-class": "org.apache.spark.sql.catalyst.TableIdentifier",
                  "table": "dst_table_test_from_src_table_df",
                  "database": "test_db"
                },
                "tableType": {
                  "product-class": "org.apache.spark.sql.catalyst.catalog.CatalogTableType",
                  "name": "MANAGED"
                },
                "storage": {
                  "product-class": "org.apache.spark.sql.catalyst.catalog.CatalogStorageFormat",
                  "compressed": false,
                  "properties": null
                },
                "schema": {
                  "type": "struct",
                  "fields": []
                },
                "provider": "parquet",
                "partitionColumnNames": [],
                "owner": "",
                "createTime": 1702736481797,
                "lastAccessTime": -1,
                "createVersion": "",
                "properties": null,
                "unsupportedFeatures": [],
                "tracksPartitionsInCatalog": false,
                "schemaPreservesCase": true,
                "ignoredProperties": null
              },
              "mode": null,
              "query": 0,
              "outputColumnNames": "[id, randomid, zip]"
            },
            {
              "class": "org.apache.spark.sql.execution.datasources.LogicalRelation",
              "num-children": 0,
              "relation": null,
              "output": [
                [
                  {
                    "class": "org.apache.spark.sql.catalyst.expressions.AttributeReference",
                    "num-children": 0,
                    "name": "id",
                    "dataType": "long",
                    "nullable": true,
                    "metadata": {},
                    "exprId": {
                      "product-class": "org.apache.spark.sql.catalyst.expressions.ExprId",
                      "id": 9,
                      "jvmId": "78343417-b80d-45a2-b489-e7d8920e61dd"
                    },
                    "qualifier": []
                  }
                ],
                [
                  {
                    "class": "org.apache.spark.sql.catalyst.expressions.AttributeReference",
                    "num-children": 0,
                    "name": "randomid",
                    "dataType": "string",
                    "nullable": true,
                    "metadata": {},
                    "exprId": {
                      "product-class": "org.apache.spark.sql.catalyst.expressions.ExprId",
                      "id": 10,
                      "jvmId": "78343417-b80d-45a2-b489-e7d8920e61dd"
                    },
                    "qualifier": []
                  }
                ],
                [
                  {
                    "class": "org.apache.spark.sql.catalyst.expressions.AttributeReference",
                    "num-children": 0,
                    "name": "zip",
                    "dataType": "string",
                    "nullable": true,
                    "metadata": {},
                    "exprId": {
                      "product-class": "org.apache.spark.sql.catalyst.expressions.ExprId",
                      "id": 11,
                      "jvmId": "78343417-b80d-45a2-b489-e7d8920e61dd"
                    },
                    "qualifier": []
                  }
                ]
              ],
              "catalogTable": {
                "product-class": "org.apache.spark.sql.catalyst.catalog.CatalogTable",
                "identifier": {
                  "product-class": "org.apache.spark.sql.catalyst.TableIdentifier",
                  "table": "src_table_test",
                  "database": "test_db"
                },
                "tableType": {
                  "product-class": "org.apache.spark.sql.catalyst.catalog.CatalogTableType",
                  "name": "MANAGED"
                },
                "storage": {
                  "product-class": "org.apache.spark.sql.catalyst.catalog.CatalogStorageFormat",
                  "locationUri": null,
                  "inputFormat": "org.apache.hadoop.hive.ql.io.parquet.MapredParquetInputFormat",
                  "outputFormat": "org.apache.hadoop.hive.ql.io.parquet.MapredParquetOutputFormat",
                  "serde": "org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe",
                  "compressed": false,
                  "properties": null
                },
                "schema": {
                  "type": "struct",
                  "fields": [
                    {
                      "name": "id",
                      "type": "long",
                      "nullable": true,
                      "metadata": {}
                    },
                    {
                      "name": "randomid",
                      "type": "string",
                      "nullable": true,
                      "metadata": {}
                    },
                    {
                      "name": "zip",
                      "type": "string",
                      "nullable": true,
                      "metadata": {
                        "__CHAR_VARCHAR_TYPE_STRING": "varchar(10)"
                      }
                    }
                  ]
                },
                "provider": "parquet",
                "partitionColumnNames": [],
                "owner": "test_user@ad.test.net",
                "createTime": 1682523315000,
                "lastAccessTime": 0,
                "createVersion": "3.3.1",
                "properties": null,
                "stats": null,
                "unsupportedFeatures": [],
                "tracksPartitionsInCatalog": false,
                "schemaPreservesCase": true,
                "ignoredProperties": null
              },
              "isStreaming": false
            }
          ]
        },
        "spark_version": {
          "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
          "_schemaURL": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/RunFacet",
          "spark-version": "3.3.1",
          "openlineage-spark-version": "1.5.0"
        },
        "processing_engine": {
          "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
          "_schemaURL": "https://openlineage.io/spec/facets/1-1-0/ProcessingEngineRunFacet.json#/$defs/ProcessingEngineRunFacet",
          "version": "3.3.1",
          "name": "spark",
          "openlineageAdapterVersion": "1.5.0"
        },
        "environment-properties": {
          "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
          "_schemaURL": "https://openlineage.io/spec/2-0-2/OpenLineage.json#/$defs/RunFacet",
          "environment-properties": {}
        }
      }
    },
    "job": {
      "namespace": "TESTSchedulerID",
      "name": "test_user_spark.execute_create_data_source_table_as_select_command.dst_table_test_from_src_table_df",
      "facets": {}
    },
    "inputs": [
      {
        "namespace": "s3a://test_db-db",
        "name": "src_table_test",
        "facets": {
          "dataSource": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/DatasourceDatasetFacet.json#/$defs/DatasourceDatasetFacet",
            "name": "s3a://test_db-db",
            "uri": "s3a://test_db-db"
          },
          "schema": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/SchemaDatasetFacet.json#/$defs/SchemaDatasetFacet",
            "fields": [
              {
                "name": "id",
                "type": "long"
              },
              {
                "name": "randomid",
                "type": "string"
              },
              {
                "name": "zip",
                "type": "string"
              }
            ]
          },
          "symlinks": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/SymlinksDatasetFacet.json#/$defs/SymlinksDatasetFacet",
            "identifiers": [
              {
                "namespace": "hive://hive-metastore.hive.svc.cluster.local:9083",
                "name": "shopify.raw_product_catalog",
                "type": "TABLE"
              }
            ]
          }
        },
        "inputFacets": {}
      }
    ],
    "outputs": [
      {
        "namespace": "s3a://test_db-db",
        "name": "dst_table_test_from_src_table_df",
        "facets": {
          "dataSource": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/DatasourceDatasetFacet.json#/$defs/DatasourceDatasetFacet",
            "name": "s3a://test_db-db",
            "uri": "s3a://test_db-db"
          },
          "schema": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/SchemaDatasetFacet.json#/$defs/SchemaDatasetFacet",
            "fields": [
              {
                "name": "id",
                "type": "long"
              },
              {
                "name": "randomid",
                "type": "string"
              },
              {
                "name": "zip",
                "type": "string"
              }
            ]
          },
          "columnLineage": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-1/ColumnLineageDatasetFacet.json#/$defs/ColumnLineageDatasetFacet",
            "fields": {
              "id": {
                "inputFields": [
                  {
                    "namespace": "s3a://test_db-db",
                    "name": "/src_table_test",
                    "field": "comments"
                  }
                ]
              },
              "randomid": {
                "inputFields": [
                  {
                    "namespace": "s3a://test_db-db",
                    "name": "/src_table_test",
                    "field": "products"
                  }
                ]
              },
              "zip": {
                "inputFields": [
                  {
                    "namespace": "s3a://test_db-db",
                    "name": "/src_table_test",
                    "field": "platform"
                  }
                ]
              }
            }
          },
          "symlinks": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/SymlinksDatasetFacet.json#/$defs/SymlinksDatasetFacet",
            "identifiers": [
              {
                "namespace": "hive://hive-metastore.hive.svc.cluster.local:9083",
                "name": "shopify.fact_order_new5",
                "type": "TABLE"
              }
            ]
          },
          "lifecycleStateChange": {
            "_producer": "https://github.com/OpenLineage/OpenLineage/tree/1.5.0/integration/spark",
            "_schemaURL": "https://openlineage.io/spec/facets/1-0-0/LifecycleStateChangeDatasetFacet.json#/$defs/LifecycleStateChangeDatasetFacet",
            "lifecycleStateChange": "CREATE"
          }
        },
        "outputFacets": {}
      }
    ]
  }
"""

_data = loads(_data)

a = RunEvent(**_data)

print(DeepDiff(_data, loads(a.json())))
```

Reply:
> What I am missing to test this is events causing issues that you've mentioned @JDarDagran

Now I'm testing it back again and it does contain additional properties. I must have missed that before.

However, you're getting rid of `_schemaURL` fields. All facets are deserialized into the base ones (e.g. 
`LifecycleStateChangeDatasetFacet` becomes `DatasetFacet), therefore they would lose schema URLs for the specific facets.

If we allow `_schemaURL` and `_producer` be arguments then probably it might work, even if the structure of the deserialized event differs.

Reply:
> @dolfinus could you provide small example with minimal set of classes just to show how you would see this?

<details>

```python
from typing import TypedDict
from uuid import UUID

from pydantic import BaseModel, ConfigDict, field_validator


class BaseFacet(BaseModel):
  # allow setting custom object attributes
  model_config = ConfigDict(extra="allow")

class RunFacet(BaseFacet):
  pass

class ParentRunFacet(RunFacet):
  pass

class ProcessingEngineRunFacet(RunFacet):
  version: str

# total=False to make all keys optional
class RunFacetsDict(TypedDict, total=False):
  parent: ParentRunFacet
  processing_engine: ProcessingEngineRunFacet

# replace with after https://peps.python.org/pep-0728/:
# class RunFacetsDict(TypedDict, total=False, closed=False):
#   parent: ParentRunFacet
#   processing_engine: ProcessingEngineRunFacet
#   __extra_items__: RunFacet
RunFacetsDict.__pydantic_config__ = ConfigDict(extra="allow")

class Run(BaseModel):
  runId: UUID
  facets: RunFacetsDict

  @field_validator("facets")
  @classmethod
  def _convert_unknown_facets(cls, facets):
    """Convert all custom facets to RunFacet, instead of nested dicts"""
    for key, value in facets.items():
      if not isinstance(value, RunFacet):
        facets[key] = RunFacet.parse_obj(value)
    return facets




class DatasetFacet(BaseFacet):
  _deleted: bool = False

class SchemaDatasetFacet(DatasetFacet):
    fields: list[dict]

class ColumnLineageDatasetFacet(DatasetFacet):
    fields: dict

# total=False to make all keys optional
class DatasetFacetsDict(TypedDict, total=False):
  schema: SchemaDatasetFacet
  columnLineage: ColumnLineageDatasetFacet

DatasetFacetsDict.__pydantic_config__ = ConfigDict(extra="allow")

class Dataset(BaseModel):
  namespace: str
  name: str
  facets: DatasetFacetsDict

  @field_validator("facets", mode="before")
  @classmethod
  def _convert_unknown_facets(cls, facets):
    """Convert all custom facets to DatasetFacet, instead of nested dicts"""
    for key, value in facets.items():
      if key not in ["schema", "columnLineage"] and not isinstance(value, DatasetFacet):
        facets[key] = DatasetFacet.parse_obj(value)
    return facets



run = Run(
  runId=UUID("123e4567-e89b-12d3-a456-426655440000"),
  facets={
    "parent": {
      "runId": "123e4567-e89b-12d3-a456-426655440000",
    },
    "processing_engine": {
      "version": "1.0.0",
    },
    "unknown": {"foo": "bar"},
  },
)

dataset = Dataset(
  namespace="default",
  name="table",
  facets={
    "schema": {
      "fields": [
        {
          "name": "foo",
          "type": "string",
        },
        {
          "name": "bar",
          "type": "int",
        },
      ],
    },
    "columnLineage": {
      "fields": {
          "foo": {
            "source": {
                "namespace": "default",
                "name": "table",
            },
            "target": {
                "namespace": "default",
                "name": "table",
            },
          },
        },
    },
    "unknown": {"foo": "bar"},
  },
)

print(repr(run))
print(repr(dataset))
```

```python
Run(runId=UUID('123e4567-e89b-12d3-a456-426655440000'), facets={'parent': ParentRunFacet(runId='123e4567-e89b-12d3-a456-426655440000'), 'processing_engine': ProcessingEngineRunFacet(version='1.0.0'), 'unknown': RunFacet(foo='bar')})

Dataset(namespace='default', name='table', facets={'schema': SchemaDatasetFacet(fields=[{'name': 'foo', 'type': 'string'}, {'name': 'bar', 'type': 'int'}]), 'columnLineage': ColumnLineageDatasetFacet(fields={'foo': {'source': {'namespace': 'default', 'name': 'table'}, 'target': {'namespace': 'default', 'name': 'table'}}}), 'unknown': DatasetFacet(foo='bar')})
```

</details>

Reply:
> That if statement in validator looks ugly honestly 😨
Fixed

Reply:
> I've actually lost the track. https://github.com/mgorsk1/OpenLineage/pull/1/files already introduces pydantic models, not cattrs

There's no decision we want to use Pydantic. attrs is current solution and is preferred way to go. However, if pydantic turns out to be faster/cleaner/whatever other reason we might want to choose pydantic instead.

Reply:
I can create an issue with describing pydantic pros and cons

Reply:
Have you thought of having separate `server` model that would more fit the deserialization needs? I see that all the solutions here have some drawbacks (some I don't like more than others, like having to have class that lists all the possible facets), but maybe if you deserialize to something else than the client classes we use to emit events they disappear?

I think it's often used, but kinda annoying Python mental model that makes you work with the same class in dozen different contexts on different abstraction levels - like active record pattern being used at business logic layer in server code.

Reply:
> Have you thought of having separate server model that would more fit the deserialization needs?

IMHO introducing another set of models just to automatically parse `facets["sql"]` as `SQLJobFacet` is too much. If it is not easy to implement (e.g. requires custom model generator) and there are other issues with this approach, then it should be generic `facets: dict[string, RunFacet]` and `facets: dict[string, DatasetFacet]`.